src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java

```diff
@@ -0,0 +1,3394 @@
+         */
+        public RandomDoublesSpliterator(AbstractSpliteratorGenerator generatingGenerator,
+                                        long index, long fence, double origin, double bound) {
+            super(index, fence);
+            this.generatingGenerator = generatingGenerator;
+            this.origin = origin; this.bound = bound;
+        }
+
+        public Spliterator.OfDouble trySplit() {
+            long i = index, m = (i + fence) >>> 1;
+            if (m <= i) return null;
+            index = m;
+            // The same generatingGenerator is used, with no splitting or copying.
+            return new RandomDoublesSpliterator(generatingGenerator, i, m, origin, bound);
+        }
+
+        public boolean tryAdvance(DoubleConsumer consumer) {
+            if (consumer == null) throw new NullPointerException();
+            long i = index, f = fence;
+            if (i < f) {
+                consumer.accept(RandomSupport.boundedNextDouble(generatingGenerator, origin, bound));
+                index = i + 1;
+                return true;
+            }
+            else return false;
+        }
+
+        public void forEachRemaining(DoubleConsumer consumer) {
+            if (consumer == null) throw new NullPointerException();
+            long i = index, f = fence;
+            if (i < f) {
+                index = f;
+                RandomGenerator r = generatingGenerator;
+                double o = origin, b = bound;
+                do {
+                    consumer.accept(RandomSupport.boundedNextDouble(r, o, b));
+                } while (++i < f);
+            }
+        }
+    }
+
+    /**
+     * Implementation support for the {@code nextExponential} method of
+     * {@link java.util.random.RandomGenerator}.
+     *
+     * Certain details of the algorithm used in this method may depend critically
+     * on the quality of the low-order bits delivered by {@code nextLong()}.  This method
+     * should not be used with RNG algorithms (such as a simple Linear Congruential
+     * Generator) whose low-order output bits do not have good statistical quality.
+     *
+     * @implNote The reference implementation uses McFarland's fast modified
+     * ziggurat algorithm (largely table-driven, with rare cases handled by
+     * computation and rejection sampling). Walker's alias method for sampling
+     * a discrete distribution also plays a role.
+     *
+     * @param rng an instance of {@code RandomGenerator}, used to generate uniformly
+     *            (pseudo)randomly chosen {@code long} values
+     *
+     * @return a nonnegative {@code double} value chosen (pseudo)randomly
+     *         from an exponential distribution whose mean is 1
+     */
+    public static double computeNextExponential(RandomGenerator rng) {
+        /*
+         * The tables themselves, as well as a number of associated parameters, are
+         * defined in class java.util.DoubleZigguratTables, which is automatically
```

Comment:
`DoubleZigguratTables` is an inner class of `RandomSupport`

Reply:
Late change fixing.